__tests__/component/form/EditCustomerForm.test.tsx

```diff
@@ -0,0 +1,142 @@
+    await waitFor(() => fireEvent.click(submitButton))
+}
+
+test("a customer can be edited with required fields", async () => {
+    expect(testCustomer.id).toBeDefined()
+    render(
+        <>
+            {testCustomer.id && (
+                <EditCustomerForm customer={thirdTestCustomer} />
+            )}
+        </>
+    )
+    await fillAndSubmitForm(testCustomerRequiredFields)
+
+    await waitFor(() => expect(bodySpy.callCount).toEqual(1))
+    expect(pathSpy.getCalls()[0].args[0]).toStrictEqual(customerEndpointURL)
+    expect(bodySpy.getCalls()[0].args[0]).toStrictEqual({
+        id: testCustomer.id,
+        ...testCustomerRequiredFields,
+    })
+})
+
+test("a customer can be edited with all fields", async () => {
+    expect(testCustomer.id).toBeDefined()
+    render(
+        <>{testCustomer.id && <EditCustomerForm customer={testCustomer} />}</>
+    )
+    await fillAndSubmitForm(testCustomerAllFields)
+
+    await waitFor(() => expect(bodySpy.callCount).toEqual(1))
+    expect(pathSpy.getCalls()[0].args[0]).toStrictEqual(customerEndpointURL)
+    expect(bodySpy.getCalls()[0].args[0]).toStrictEqual({
+        id: testCustomer.id,
+        ...testCustomerAllFields,
+    })
+})
+
+test("afterSubmit is invoked with the correct data", async () => {
+    const afterSubmitSpy = spy((createTaskResponse) => createTaskResponse)
+    expect(testCustomer.id).toBeDefined()
+    render(
+        <>
+            {testCustomer.id && (
+                <EditCustomerForm
+                    afterSubmit={afterSubmitSpy}
+                    customer={testCustomer}
+                />
+            )}
+        </>
+    )
+    await fillAndSubmitForm(testCustomerAllFields)
+
+    await waitFor(() => expect(afterSubmitSpy.callCount).toEqual(1))
+    expect(afterSubmitSpy.getCalls()[0].args[0]).toStrictEqual({
+        isSuccess: true,
+        isError: false,
+        data: { id: testCustomer.id, ...testCustomerAllFields },
+    })
+})
+
+test("onCancel is invoked", async () => {
+    const onCancelSpy = spy(() => null)
+    render(<EditCustomerForm onCancel={onCancelSpy} customer={testCustomer} />)
+
+    const cancelButton = screen.getByTestId("form-button-cancel")
+    await waitFor(() => fireEvent.click(cancelButton))
+
+    await waitFor(() => expect(onCancelSpy.callCount).toEqual(1))
+})
+
+test("a required field cannot be missing", async () => {
+    const submitTimeout = 100
+    for (const field of Object.keys(testCustomerRequiredFields)) {
+        const form = render(<EditCustomerForm customer={testCustomer} />)
+        const customerMissingRequired = _.omit(testCustomerAllFields, field)
+
+        /* eslint-disable no-await-in-loop */
+        await fillAndSubmitForm(customerMissingRequired as Customer)
+        await new Promise((resolve) =>
+            setTimeout(() => resolve(null), submitTimeout)
+        )
+        /* eslint-enable */
+
+        expect(pathSpy.callCount).toEqual(1)
```

Comment:
This is incorrect. This line asserts that the form HAS been sent, while it should assert that the form has NOT been sent (since a required field is missing). The correct line is
```
expect(pathSpy.callCount).toEqual(0)
```
The form and the tests should work so that this is true.

I'll take a look and get back to you about this.